doc/sphinx/proof-engine/ltac.rst

```diff
@@ -469,45 +500,117 @@ We can check if a tactic made progress with:
-
-We can branch with the following structure:
+Branching and backtracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. tacn:: @ltac_expr__1 + @ltac_expr__2
+|Ltac| provides several :gdef:`branching` tactics that permit trying multiple alternative tactics
+for a proof step.  For example, :tacn:`first`, which tries several alternatives and selects the first
+that succeeds, or :tacn:`tryif`, which tests whether a given tactic would succeed or fail if it was
+applied and then, depending on the result, applies one of two alternative tactics.  There are
+are also looping constructs :tacn:`do` and :tacn:`repeat`.  The order in which the subparts
+of these tactics are evaluated is generally similar to
+structured programming constructs in many languages.
+
+The :tacn:`+<+ (backtracking branching)>`, :tacn:`multimatch` and :tacn:`multimatch goal` tactics
+provide more complex capability.  Rather than applying a single successful
+tactic, these tactics generate a series of successful tactic alternatives that are tried sequentially
+when subsequent tactics outside these constructs fail.  Please examine the :tacn:`multimatch` example
+:ref:`here<match_vs_multimatch_ex>`.
+
+These constructs are evaluated using :gdef:`backtracking`.  Each  creates a
+:gdef:`backtracking point`.  When a subsequent tactic fails, evaluation continues from the nearest
+prior backtracking point with the next successful alternative and repeats the tactics after
+the backtracking point.  When a backtracking point has
+no more successful alternatives, evaluation continues from the next prior backtracking point.
+If there are no more prior backtracking points, the overall tactic fails.
+
+Thus, backtracking tactics can have multiple successes.  Non-backtracking constructs that appear
+after a backtracking point are reprocessed after backtracking, as in the example
+:ref:`here<match_vs_multimatch_ex>`, in which
+the :tacn:`;<ltac-seq>` construct is reprocessed after backtracking.  When a backtracking construct is within
+a non-backtracking construct, the latter uses the :gdef:`first success`.  Backtracking to
+a point within a non-backtracking construct won't change the branch that was selected by the
+non-backtracking construct.
+
+The :tacn:`once` tactic stops further backtracking to backtracking points within that tactic.
+
+eval vs execution (also out of order)
+
+Branching with backtracking: +
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can branch with backtracking with the following structure:
+
+.. tacn:: @ltac_expr1 + {| @ltac_expr2 | @binder_tactic }
    :name: + (backtracking branching)
 
-   :n:`@ltac_expr__1` and :n:`@ltac_expr__2` are evaluated to :n:`v__1` and
+   :n:`@ltac_expr1` and :n:`@ltac_expr2` are evaluated to :n:`v__1` and
    :n:`v__2`  respectively, which must be tactic values. The tactic value :n:`v__1` is applied to
    each focused goal independently and if it fails or a later tactic fails, then
    the proof backtracks to the current goal and :n:`v__2` is applied.
```

Comment:
This is not technically accurate; `ltac_expr2` is only evaluated to a value after backtracking.  For example, the following only poses `I` once:
```coq
Goal True.
  (let __ := match goal with _ => pose I end in idtac)
  + (let __ := match goal with _ => pose I end in idtac).
```

Reply:
Huh?  How does that reconcile with this?

```
Goal True.
idtac "hi" + fresh "h".

(* error: Tactic expected. *)
```
```

Reply:
That is a parse error, not an evaluation error.  It also shows up in
```coq
Goal True.
  idtac "hi" + (idtac; fresh "h").
```
and, more tellingly, on
```coq
Ltac foo := idtac "hi" + fresh "h".
```
where no evaluation is done.

Reply:
I wouldn't call that a parser error.  It happens after parsing succeeds.  Furthermore the message is poorly worded.  It should say something like "This subexpression must return a tactic value".

For example:

```
Ltac foo := idtac "hi" + (fresh "h") + ?.

(* Syntax error: [tactic:binder_tactic] or [tactic:tactic_expr] expected after '+' (in [tactic:tactic_expr]). *)
```

I'll take your word on the original point here.

Reply:
You're right, it's not a parsing error, but an internalization error.  (I guess this is the step after parsing, when the parsed object turns into a tactic AST or something?)
https://github.com/coq/coq/blob/9c26e583668827bff5159e7671c406ace8b2f3ae/plugins/ltac/tacintern.ml#L35-L36
which is called only at
https://github.com/coq/coq/blob/9c26e583668827bff5159e7671c406ace8b2f3ae/plugins/ltac/tacintern.ml#L689-L695